src/test/resources/com/jpinpoint/perf/lang/java/ruleset/common/xml/MDCPutWithoutRemove.xml

```diff
@@ -156,4 +156,114 @@ public class TracingInterceptor extends HandlerInterceptorAdapter {
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import org.slf4j.MDC;
+import org.springframework.security.web.server.authentication.ServerAuthenticationConverter;
+import reactor.core.publisher.Mono;
+import org.springframework.web.server.ServerWebExchange;
+
+class Issue326 {
+        private static final String MDC_REQUEST_ID_KEY = "Request-id";
+        private final ServerAuthenticationConverter authenticationConverter;
+
+       public Mono<Void> filter(ServerWebExchange serverWebExchange) {
+            MDC.put(MDC_REQUEST_ID_KEY, "something"); // <--- false positive
+            return authenticationConverter.convert(serverWebExchange)
+                   .doFinally(signalType -> MDC.remove(MDC_REQUEST_ID_KEY));
+        }
+        public Mono<Void> filter2(ServerWebExchange serverWebExchange) {
+            MDC.put(MDC_REQUEST_ID_KEY, "something"); // <--- false positive
+            return authenticationConverter.convert(serverWebExchange)
+                   .doFinally(signalType -> MDC.clear());
+        }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>3 violations: Issue 473: put in called method</description>
+        <expected-problems>3</expected-problems>
+        <expected-linenumbers>29,59,73</expected-linenumbers>
+        <code><![CDATA[
+import org.slf4j.MDC;
+
+class Issue473_a {
+    public void foo() {
+        try {
+            MDC.put("a-key", "a-value"); // ok
+            bar();
+        } finally {
+            MDC.remove("a-key");
+            MDC.remove("b-key");
+        }
+    }
+    private void bar() {
+        MDC.put("b-key", "b-value"); // <--- false positive
+    }
+}
+
+class Issue473_b {
+    public void foo() {
+        try {
+            MDC.put("a-key", "a-value"); // ok
+            bar();
+        } finally {
+            MDC.remove("a-key");
+            MDC.remove("b-key");
+        }
+    }
+    private void bar() {
+        MDC.put("b-key", "b-value"); // bad - because called from non-try-finally place
+    }
+    void baz() {
+        bar();
+    }
+}
+class Issue473_c {
+    public void foo() {
+        try {
+            MDC.put("a-key", "a-value"); // ok
+            bar();
+        } finally {
+            MDC.clear();
+        }
+    }
+    private void bar() {
+        MDC.put("b-key", "b-value"); // <--- false positive
+    }
+}
+class Issue473_d {
+    public void foo() {
+        try {
+            MDC.put("a-key", "a-value"); // ok
+            bar();
+        } finally {
+            MDC.remove("a-key");
+        }
+    }
+
+    private void bar() {
+        MDC.put("b-key", "b-value"); // bad, key not removed
+    }
+}
+class Issue473_e {
+    public void foo() {
+        try {
+            MDC.put("a-key", "a-value"); // ok
+            bar();
+        } finally {
+            MDC.clear();
+        }
+    }
+
+    public void bar() {
+        MDC.put("b-key", "b-value"); // bad, method public
```

Comment:
should this also be the `a-key`? even if bar() is private it still fails because the key is different?

Reply:
Not sure what you mean?

Reply:
right :-)